lib/segment/src/vector_storage/simple_multi_dense_vector_storage.rs

```diff
@@ -0,0 +1,212 @@
+use std::ops::Range;
+use std::sync::atomic::AtomicBool;
+use std::sync::Arc;
+
+use atomic_refcell::AtomicRefCell;
+use bitvec::prelude::{BitSlice, BitVec};
+use common::types::PointOffsetType;
+use parking_lot::RwLock;
+use rocksdb::DB;
+
+use crate::common::operation_error::{check_process_stopped, OperationError, OperationResult};
+use crate::common::rocksdb_wrapper::DatabaseColumnWrapper;
+use crate::common::Flusher;
+use crate::data_types::named_vectors::CowVector;
+use crate::data_types::vectors::{DenseVector, MultiDenseVector, VectorRef};
+use crate::types::Distance;
+use crate::vector_storage::bitvec::bitvec_set_deleted;
+use crate::vector_storage::common::StoredRecord;
+use crate::vector_storage::{VectorStorage, VectorStorageEnum};
+
+type StoredMultiDenseVector = StoredRecord<MultiDenseVector>;
+
+/// In-memory vector storage with on-update persistence using `store`
+#[allow(unused)]
+pub struct SimpleMultiDenseVectorStorage {
+    dim: usize,
+    distance: Distance,
+    /// Keep vectors in memory
+    vectors: Vec<MultiDenseVector>,
+    db_wrapper: DatabaseColumnWrapper,
+    update_buffer: StoredMultiDenseVector,
+    /// BitVec for deleted flags. Grows dynamically upto last set flag.
+    deleted: BitVec,
+    /// Current number of deleted vectors.
+    deleted_count: usize,
+}
+
+#[allow(unused)]
+pub fn open_simple_multi_dense_vector_storage(
+    database: Arc<RwLock<DB>>,
+    database_column_name: &str,
+    dim: usize,
+    distance: Distance,
+    stopped: &AtomicBool,
+) -> OperationResult<Arc<AtomicRefCell<VectorStorageEnum>>> {
+    let mut vectors = vec![];
+    let (mut deleted, mut deleted_count) = (BitVec::new(), 0);
+    let db_wrapper = DatabaseColumnWrapper::new(database, database_column_name);
+
+    let mut total_vector_count = 0;
+    let mut total_sparse_size = 0;
+    db_wrapper.lock_db().iter()?;
+    for (key, value) in db_wrapper.lock_db().iter()? {
+        let point_id: PointOffsetType = bincode::deserialize(&key)
+            .map_err(|_| OperationError::service_error("cannot deserialize point id from db"))?;
+        let stored_record: StoredMultiDenseVector = bincode::deserialize(&value)
+            .map_err(|_| OperationError::service_error("cannot deserialize record from db"))?;
+
+        // Propagate deleted flag
+        if stored_record.deleted {
+            bitvec_set_deleted(&mut deleted, point_id, true);
+            deleted_count += 1;
+        }
+        vectors.insert(point_id as usize, stored_record.vector);
```

Comment:
`insert` panics when point_id >= length
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.insert
also it shifts elements